File: DailyChallenges/Scala/LeetCode_103.scala
object Solution {
    def zigzagLevelOrder(root: TreeNode): List[List[Int]] = {
        val q = collection.mutable.Queue.empty[TreeNode]; q.enqueue(root)
        var ans: List[List[Int]] = Nil
        while (q.nonEmpty){
            var l: List[Int] = Nil
            for (i<- 0 until q.length; node = q.dequeue if node != null){
                l = node.value :: l
                q.enqueue(node.left)
                q.enqueue(node.right)
            }
            if (l.nonEmpty)
                ans = (if (ans.length % 2 == 0) l.reverse else l) :: ans 
        }
        ans.reverse
    }
}

File: DailyChallenges/CPP/LeetCode_103.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        queue<TreeNode *> q; q.push(root);
        vector<vector<int>> ans;
        while(!q.empty()){
            vector<int> v;
            int sz = q.size();
            for (int i = 0; i<sz; ++i){
                auto node = q.front(); q.pop();
                if (node){
                    v.push_back(node->val);
                    q.push(node->left);
                    q.push(node->right);
                }
            }
            if (ans.size() % 2 == 1)
                reverse(v.begin(), v.end());
            if (!v.empty())
                ans.push_back(v);
        }
        return ans;
    }
};
